mirrord/layer/src/file/hooks.rs

```diff
@@ -393,6 +394,54 @@ unsafe fn fileno_logic(file_stream: *mut FILE) -> c_int {
     }
 }
 
+/// Tries to convert input to type O, if it fails it returns the max value of O.
+/// For example, if you put u32::MAX into a u8, it will return u8::MAX.
+fn best_effort_cast<I: Bounded, O: TryFrom<I> + Bounded>(input: I) -> O {
+    input.try_into().unwrap_or_else(|_| O::max_value())
+}
+
+/// Converts time in nano seconds to seconds, to match the `stat` struct
+/// which has very weird types used
+fn nano_to_secs(nano: i64) -> i64 {
+    best_effort_cast(Duration::from_nanos(best_effort_cast(nano)).as_secs())
+}
+
+/// Hook for `libc::lstat`.
+#[hook_guard_fn]
+unsafe extern "C" fn lstat_detour(raw_path: *const c_char, out_stat: *mut stat) -> c_int {
+    let path = (!raw_path.is_null()).then(|| CStr::from_ptr(raw_path));
+
+    let (Ok(result) | Err(result)) = lstat(path)
+        .map(|res| {
+            let res = res.metadata;
```

Comment:
Why's all this in the hook and not in the `lstat` method?

Reply:
following how other hooks work - the `ops` methods are rusty and safe, and here we assign to a `*mut T` (raw pointer) values from the rust struct, which is unsafe so is done here.